gradle-sls-packaging/src/main/java/com/palantir/gradle/dist/BaseDistributionExtension.java

```diff
@@ -131,6 +134,30 @@ public final void setProductType(ProductType productType) {
         this.productType.set(productType);
     }
 
+    public final SetProperty<ArtifactLocator> getArtifacts() {
+        return artifacts;
+    }
+
+    public final void artifact(ArtifactLocator artifactLocator) {
+        artifactLocator.isValid();
+        artifacts.add(artifactLocator);
+    }
+
+    /** Lazily configures and adds a {@link ArtifactLocator}. */
+    public final void artifact(@DelegatesTo(ArtifactLocator.class) Closure<ArtifactLocator> closure) {
+        artifacts.add(providerFactory.provider(() -> {
```

Comment:
this simplifies to just:

### create the `ArtifactLocator`
You will need to do something like `project.getObjects().newInstance(ArtifactLocator.class)` to make Gradle instantiate and "managed" your managed type for you.

### configure it
can just do your `project.configure(artifactLocator, closure)`, and you don't have to worry about wrapping that in a provider since the underlying type is lazy

### just add it
add it to the domain object set, and the validation in the `configureEach` or `whenObjectAdded` will take over.

Reply:
you should also add a `artifact(Action<ArtifactLocator> action) { ... } ` method. That gives you a type safe way of adding it to the set when using Java. Buildscript will likely pick the `Closure` form.

Instead of `project.configure(...)` you end up just doing `action.execute(locator)`.